### ISSUE TITLE
Add user-defined conversion operators for C#

### ISSUE THREAD

**Issue by emmenlau** [swig/swig#2367]
I've just started with SWIG and followed the documentation to build a few first examples. Everything works great!

However a question: I can not use directly the native C# type in a method signature that expects the generated type. It seems this is because the managed classes are missing a [user-defined conversion operator](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/user-defined-conversion-operators).

This code illustrates the problem in C#, for an interface that has a `StringVector` name for `std::vector<std::string>`:
```
    string[] cData = { "Hello", "World", "äöü", "ß 是" };
    IList<string> vStringList = new List<string>(cData);

    // This does not compile, with error:
    // Cannot convert from 'System.Collections.Generic.IList<string>' to 'StringVector' [InteropTestManaged]
    StringVector vStringVector = vStringList;

    // Due to this problem, I can not use vStringList in the interface:
    // This does not compile:
    //vInterface.setStringVector(vStringList)
```

For reference, this was originally asked on the mailling list at https://sourceforge.net/p/swig/mailman/message/37706110/

**Comment by wsfulton** [comment]
You can use `vInterface.setStringVector(new StringVector(vStringList))` instead of `vInterface.setStringVector(vStringList)` for now. If we add user-defined conversion operators, I would like to know which version of .NET this is supported by as we may have to adjust the minimum supported version.

**Comment by emmenlau** [comment]
> You can use `vInterface.setStringVector(new StringVector(vStringList))` instead of `vInterface.setStringVector(vStringList)` for now.

Yes, for `std::vector` that works, thanks. However I soon found that other types like `std::map` do not support this syntax(?)

> If we add user-defined conversion operators, I would like to know which version of .NET this is supported by as we may have to adjust the minimum supported version.

I could not find a good reference for which version adds the feature, but I found this website: https://www.codeproject.com/Tips/452213/Forgotten-Csharp-language-features-implicit-operat that suggests the feature had been available since .Net 1.1.

**Comment by emmenlau** [comment]
A demo of user-defined conversion operators for `std::vector` and `std::map` is in #2414
With these operators, its possible to write in C#:
```
Dictionary<string, string> vDict;
// StringStringMap is a SWIG-generated binding for std::map<std::string, std::string>:
StringStringMap vMap = vDict;
```

**Comment by emmenlau** [comment]
I've used the conversion operators now for a bit, and I'm quite happy with them. Could you kindly consider if this goes in the right direction, or advise how to proceed?

**Comment by wsfulton** [comment]
Hi @emmenlau, looks okay with a cursory look. Pull requests require tests before being accepted. Can you please enhance one of the runtime testcases that tests the new functionality. Thanks.
